plugins/org.eclipse.n4js.flowgraphs/src/org/eclipse/n4js/flowgraphs/factories/BinaryLogicalExpressionFactory.java

```diff
@@ -53,24 +64,65 @@ static ComplexNode buildComplexNode(ReentrantASTIterator astpp, BinaryLogicalExp
 			break;
 		}
 
-		cNode.connectInternalSucc(entryNode, lhsNode);
-		cNode.connectInternalSucc(thenCFT, lhsNode, rhsNode);
+		cNode.connectInternalSucc(entryNode, lhsNode, scJumpNode);
+		cNode.connectInternalSucc(thenCFT, scJumpNode, rhsNode);
 		cNode.connectInternalSucc(rhsNode, exitNode);
 
-		cNode.connectInternalSucc(elseCFT, lhsNode, exitNode); // short-circuit evaluation
-
-		if (lhsNode == null) { // broken AST
-			if (rhsNode == null) {
-				cNode.connectInternalSucc(entryNode, exitNode);
-			} else {
-				cNode.connectInternalSucc(entryNode, rhsNode);
-			}
-		}
+		scJumpNode.addJumpToken(new JumpToken(elseCFT)); // short-circuit evaluation
+		cNode.setJumpNode(scJumpNode);
 
 		cNode.setEntryNode(entryNode);
 		cNode.setExitNode(exitNode);
 
+		rhsNode.addCatchToken(new CatchToken(thenCFT));
+
+		boolean isCatchingLhs = isTopJumpCatcher(lbExpr);
+		if (isCatchingLhs) {
+			exitNode.addCatchToken(new CatchToken(elseCFT));
+			exitNode.addCatchToken(new CatchToken(thenCFT));
+		} else {
+			// TODO: minor improvement: add the following jump node
+			// exitNode.addJumpToken(new JumpToken(thenCFT)); // short-circuit evaluation
+			// cNode.setJumpNode(scJumpNode);
```

Comment:
what's up with this `TODO`?

Reply:
It's more a reminder for an improvement that improves the concept. Indirectly, the issue itself is already solved.